BCODB submission form

HadleyKing opened the pull request:
Updated to add form for submission of BCO to BCODB

Changes to be committed:
- new file:   client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue
- modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationState.vue

See https://gitter.im/galaxyproject/dev?at=60c37cd0b346e9618bc6e67b

The above form will submit the "proto" BCO to the supplied url. Obtaining a user api key requires account creation on the BCODB portal (currently only available for [BETA](https://beta.portal.aws.biochemistry.gwu.edu/login))

## How to test the changes?
(Select all options that apply)
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Obtaining a user api key requires account creation on the BCODB portal. At [BETA.portal](https://beta.portal.aws.biochemistry.gwu.edu/) create an account
  2. Once Email verification is complete navigate to the [Account Page](https://beta.portal.aws.biochemistry.gwu.edu/account) and scroll to bottom. Copy the API token listed for the BCO Server you want to deposit a BCO to
  
![Screen Shot 2021-06-15 at 2 51 18 PM](https://user-images.githubusercontent.com/20343214/122107740-6590e200-cde9-11eb-8d25-aba0255094a2.png)

  3. In the Galaxy `Invocation` menu expand the desired invocation
  ![Screen Shot 2021-06-15 at 2 39 23 PM](https://user-images.githubusercontent.com/20343214/122106090-7b9da300-cde7-11eb-9109-d40e0c726971.png)
  4. Expanding the `BioCompute Object` item will display the previously implemented `Download BioCompute Object` and a new form that can be expanded to submit above it: `Export BioCompute Object`
  <img width="857" alt="Screen Shot 2021-06-13 at 6 59 53 PM" src="https://user-images.githubusercontent.com/20343214/121825501-cc46bc00-cc80-11eb-942d-38fcc48779bb.png">
  5. The above form will submit the BCO to the supplied url. 
<img width="821" alt="Screen Shot 2021-06-13 at 7 30 27 PM" src="https://user-images.githubusercontent.com/20343214/121825513-df598c00-cc80-11eb-9e6f-2266a3996dd8.png">

## License

- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

@nsoranzo

HadleyKing committed:
Revert "Add Job metrics for cost report"

This reverts commit 05b7e89e4651b48fd6a16079d8e6ba1f850f0c61.

HadleyKing committed:
Added functions to extract dictionary for AWS cost estimate

RE #28 and https://github.com/galaxyproject/galaxy/issues/9046 and https://github.com/galaxyproject/galaxy/pull/9621

    GET /api/invocations/{invocations_id}/aws_estimate
    Return a dictionary with an AWS estimate for a workflow invocation.
Changes to be committed:
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py
	modified:   lib/galaxy/webapps/galaxy/buildapp.py

HadleyKing committed:
Updates to address review comments by @mvdbeek and @bgruening
Changes to be committed:
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py
	modified:   lib/galaxy/webapps/galaxy/buildapp.py

HadleyKing committed:
Add invocation metrics to extension for BCO
Changes to be committed:
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
BCODB form
Updated to add form for submission of BCO to BCODB
Changes to be committed:
	new file:   client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationState.vue

HadleyKing committed:
fix missing file
Changes to be committed:
	new file:   client/src/components/User/CloudAuth/model/ResourceProviders.js

HadleyKing committed:
Fix for BCO compliance
Fix to `_generate_invocation_bco` for BCO generation.
Changes to be committed:
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py

HadleyKing committed:
fix depreciated test
Changes to be committed:
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationState.test.js

HadleyKing committed:
API/Lint testing fixes
Changes to be committed:
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py
	modified:   lib/galaxy_test/base/populators.py

HadleyKing committed:
Python Linting and Prettier
Changes to be committed:
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationState.vue
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py

Nerdinacan commented on client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue:
```diff
@@ -0,0 +1,140 @@
+<template>
+    <div v-if="Object.keys(invocation).length">
+        <details class="invocation-biocompute-objects">
```
Just a minor point, but if you find yourself doing a lot of these kinds of calculations in-template, it might be easier and read better to use an appropriately named computed. 

```js
export default {

    // ....

    computed: {
        hasInvocationProps() {
            return Object.keys(this.invocation).length > 0;
        }
    },

    // ....

}
```

Nerdinacan commented on client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue:
```diff
@@ -0,0 +1,140 @@
+                                />
+                            </div>
+                            <div class="form-group">
+                                <button class="btn btn-primary">Submit</button>
```
Assuming these are mounted with the standard component mount, you should have access to localization features. It's probably a  good habit for all of us to start doing by default:
```html
Option 1: as a directive, only works if the element you slap it on is pretty simple.
<button class="btn btn-primary" v-localize>Submit</button>

Option 2: as a filter
<button class="btn btn-primary">{{ 'Submit' | localize }}</button>
````

Nerdinacan commented on client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue:
```diff
@@ -0,0 +1,140 @@
+            },
+        };
+    },
+    created: function () {
```
If you find yourself doing an ajax call in every created handler, that's a good sign that this component might be better realized as 2 components: One whose job is to display the form fields, and one whose job is to retrieve and save the data the form works with.

This will make it easier to unit test your components and your ajax functions, and is also just good separation of concerns. I think you'll see that it's easier to start adding more complex functionality in the future too, maybe field validation etc. Those can all go in the "form" part, and any data validation or save error handling can go in the "data manager" layer which lives above.

```vue
<!-- hypothetical data wrapper around your form, separates data manipulation from data rendering -->

<template>
    <!-- now, this form component would only render the static form fields and maybe do field validation -->
    <WorkflowInvocationBioCompute v-if="bco" v-model="bco" />
    <div v-else>Loading...</div>
</template>

<script>
// Importing your ajax calls from separate files will make it easier to unit test this component because mocking
// a simple imported function is trivial with jest
import { saveInvocation, getInvocation } from "./queries";

export default {
    props: {
        invocation: { required: true },
    },
    data() {
        // Store your bco temp data exterior to the form rendering
        return { bco: null };
    },
    watch: {
        bco(newVal, oldVal) {
            if (newVal !== oldVal) {
                // do something with the form edit results
                this.save(newVal);
            }
        }
    },
    created() {
        this.load();     
    },
    methods: {
        async load() {
            // Do your ajax load here and store in temp variable
            this.bco = await getInvocation(this.invocation);
        },
        async save(newVal) {
            // basically the same as your submitForm method
            return await saveInvocation(newVal);
        }
    }
}

</script>
```

HadleyKing replied:
@Nerdinacan 
I am not sure I fully understand the implementation of your comment here. Are you suggesting I create another component with the structure you outlined above?

HadleyKing replied:
@nsoranzo suggested integrating the BCO API key into the Galaxy  user preferences

HadleyKing committed:
Axios Error handling for BCO submission
Changes to be committed:
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationState.vue

HadleyKing committed:
Fixes for js_lint and mypy
Changes to be committed:
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationState.vue
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py

HadleyKing committed:
Merge pull request #56 from galaxyproject/dev

Update from Dev

HadleyKing committed:
Merge branch 'galaxyproject:dev' into main

HadleyKing committed:
Add response messaging and clear form after submit
Changes to be committed:
	modified:   client/src/components/WorkflowInvocationState/WorkflowInvocationBiocompute.vue
